File: tools/tracer/src/tracer/commands/status.py
"""`trace status` — working-tree dirty set with code intelligence.

Replaces the `git status --short` → `git diff --stat` → per-file `git diff`
cluster agents run before commits. One call returns every uncommitted-state
file annotated with ccn rank, downstream caller count, deploy-branch
presence, and the standard lifecycle shoulder — ordered by blast radius so
the most load-bearing dirty files surface first.
"""

from __future__ import annotations

import json
from pathlib import Path

import click

from tracer import architecture, cache, file_facts, git_activity, passive_context
from tracer.deps import require_dependencies


# Order the working-state buckets so the human-readable output groups
# semantically: brand-new code first, then renames, modifications, deletions,
# then untracked at the bottom (most likely intentional, least urgent).
_STATE_ORDER = ("added", "renamed", "modified", "deleted", "untracked")


def _graph_counts(relative: str, graph) -> dict | None:
    """Caller/dependent counts for the file's module, or None if not in graph."""
    if graph is None:
        return None
    module_id = graph.file_to_module_id.get(relative)
    if not module_id:
        return None
    return {
        "callers": len(architecture.dependents_of(graph, module_id)),
        "depended_on_by_modules": len(architecture.dependencies_of(graph, module_id)),
    }


def _entries_for_state(
    repo_root: Path,
    states: dict[str, str],
    graph,
) -> list[dict]:
    """Build one annotated entry per dirty file."""
    entries: list[dict] = []
    for relative, state in states.items():
        abs_path = repo_root / relative
        facts = file_facts.get(abs_path, repo_root=repo_root) if abs_path.exists() else None
        gc = _graph_counts(relative, graph) if facts else None
        shoulder = passive_context.render(facts, graph=gc) if facts else None
        entry = {
            "path": relative,
            "state": state,
            "shoulder": shoulder,
            "callers": (gc["callers"] if gc else 0),
            "depended_on_by_modules": (gc["depended_on_by_modules"] if gc else 0),
            "ccn_total": (facts.cyclomatic_complexity_total if facts else 0),
            "ccn_rank": (facts.rank if facts else "unknown"),
            "present_in": list(facts.present_in) if facts else [],
            "last_subject": facts.last_subject if facts else None,
            "top_author": facts.top_author if facts else None,
        }
        entries.append(entry)
    return entries


def _sort_key(entry: dict) -> tuple:
    """Sort by blast radius first, complexity second, then path.

    Files with many downstream callers are load-bearing; show first.
    Negate counts so higher values sort earlier under default ascending sort.
    """
    state_rank = _STATE_ORDER.index(entry["state"]) if entry["state"] in _STATE_ORDER else len(_STATE_ORDER)
    return (
        -entry["callers"],
        -entry["ccn_total"],
        state_rank,
        entry["path"],
    )


@click.command()
@click.option("--json", "as_json", is_flag=True, help="Machine-parseable JSON output")
@click.option(
    "--state",
    "state_filter",
    type=click.Choice(_STATE_ORDER, case_sensitive=False),
    default=None,
    help="Restrict to a single working state",
)
def command(as_json: bool, state_filter: str | None) -> None:
    """Repo-wide dirty file set with code intelligence."""
    require_dependencies()
    repo_root = cache.repo_root_for(".")
    states = git_activity._working_tree_state(repo_root)
    if state_filter:
        states = {p: s for p, s in states.items() if s == state_filter}

    graph = architecture.load_cached(repo_root)
    entries = _entries_for_state(repo_root, states, graph)
    entries.sort(key=_sort_key)

    if as_json:
        click.echo(
            json.dumps(
                {
                    "repo_root": str(repo_root),
                    "count": len(entries),
                    "entries": entries,
                },
                indent=2,
            )
        )
        return

    if not entries:
        click.echo("(working tree clean)")
        return

    click.echo(f"{len(entries)} files with uncommitted state:")
    click.echo("")
    current_state: str | None = None
    for entry in entries:
        if entry["state"] != current_state:
            current_state = entry["state"]
            click.echo(f"## {current_state}")
        path = entry["path"]
        callers = entry["callers"]
        ccn = entry["ccn_total"]
        rank = entry["ccn_rank"]
        shoulder = entry["shoulder"] or ""
        suffix = f"  (callers={callers}, ccn={ccn} {rank})"
        click.echo(f"  {path}{suffix}")
        if shoulder:
            click.echo(f"    {shoulder}")

File: packages/claude/bin/tracer-dist/crate/src/commands/status.rs
//! `trace status` — working-tree dirty set with code intelligence.
//! Every uncommitted-state file annotated with ccn rank, downstream
//! caller count, deploy-branch presence, and the lifecycle shoulder,
//! ordered by blast radius. CCN is AST-derived.

use crate::{architecture, cache, file_facts, git_activity, passive_context};
use anyhow::Result;
use serde_json::{json, Value};
use std::path::Path;

/// Human-output grouping order; also the tertiary sort key.
const STATE_ORDER: &[&str] = &["added", "renamed", "modified", "deleted", "untracked"];

fn state_rank(state: &str) -> usize {
    STATE_ORDER.iter().position(|s| *s == state).unwrap_or(STATE_ORDER.len())
}

/// Caller/dependency counts for a file from the graph. None when the file
/// has no graph node.
fn graph_counts(relative: &str, graph: Option<&architecture::Graph>) -> Option<Value> {
    let graph = graph?;
    let module_id = graph.file_to_module_id.get(relative)?;
    Some(json!({
        "callers": architecture::dependents_of(graph, module_id).len(),
        "depended_on_by_modules": architecture::dependencies_of(graph, module_id).len(),
    }))
}

fn entries_for_state(
    repo_root: &Path,
    states: &[(String, String)],
    graph: Option<&architecture::Graph>,
) -> Vec<Value> {
    let mut entries = Vec::new();
    for (relative, state) in states {
        let abs_path = repo_root.join(relative);
        let facts = if abs_path.exists() {
            file_facts::get(&abs_path, repo_root, None)
        } else {
            None
        };
        let gc = if facts.is_some() {
            graph_counts(relative, graph)
        } else {
            None
        };
        let shoulder = facts
            .as_ref()
            .map(|f| passive_context::render(f, gc.as_ref()));
        entries.push(json!({
            "path": relative,
            "state": state,
            "shoulder": shoulder,
            "callers": gc.as_ref().and_then(|g| g["callers"].as_i64()).unwrap_or(0),
            "depended_on_by_modules":
                gc.as_ref().and_then(|g| g["depended_on_by_modules"].as_i64()).unwrap_or(0),
            "ccn_total":
                facts.as_ref().map(|f| f.cyclomatic_complexity_total).unwrap_or(0),
            "ccn_rank":
                facts.as_ref().map(|f| f.rank.clone()).unwrap_or_else(|| "unknown".into()),
            "present_in":
                facts.as_ref().map(|f| f.present_in.clone()).unwrap_or_default(),
            "last_subject": facts.as_ref().and_then(|f| f.last_subject.clone()),
            "top_author": facts.as_ref().and_then(|f| f.top_author.clone()),
        }));
    }
    entries
}

/// Blast-radius sort key: (-callers, -ccn_total, state_rank, path).
fn sort_key(entry: &Value) -> (i64, i64, usize, String) {
    (
        -entry["callers"].as_i64().unwrap_or(0),
        -entry["ccn_total"].as_i64().unwrap_or(0),
        state_rank(entry["state"].as_str().unwrap_or("")),
        entry["path"].as_str().unwrap_or("").to_string(),
    )
}

pub fn run(as_json: bool, state_filter: Option<&str>) -> Result<Value> {
    let here = Path::new(".");
    let repo_root = cache::worktree_root_for(here).unwrap_or_else(|| cache::display_root(here));
    let state_map = git_activity::working_tree_state(&repo_root);
    let mut states: Vec<(String, String)> = state_map
        .into_iter()
        .filter(|(_, s)| match state_filter {
            Some(f) => s == f,
            None => true,
        })
        .collect();
    // Deterministic pre-sort by path so the stable final sort resolves
    // ties by path.
    states.sort();

    // Build (not just load) the graph: `status` runs precisely when the
    // tree is dirty, and any dirty file changes the architecture
    // fingerprint, so a load-only path would miss the cache every time
    // and collapse every blast-radius count to zero.
    let graph = architecture::get(&repo_root);
    let mut entries = entries_for_state(&repo_root, &states, Some(&graph));
    entries.sort_by(|a, b| sort_key(a).cmp(&sort_key(b)));

    let value = json!({
        "repo_root": repo_root.to_string_lossy(),
        "count": entries.len(),
        "entries": entries.clone(),
    });

    if as_json {
        return Ok(value);
    }

    if entries.is_empty() {
        println!("(working tree clean)");
        return Ok(value);
    }

    println!("{} files with uncommitted state:", entries.len());
    println!();
    let mut current_state: Option<String> = None;
    for entry in &entries {
        let state = entry["state"].as_str().unwrap_or("").to_string();
        if Some(&state) != current_state.as_ref() {
            println!("## {state}");
            current_state = Some(state);
        }
        let callers = entry["callers"].as_i64().unwrap_or(0);
        let ccn = entry["ccn_total"].as_i64().unwrap_or(0);
        let rank = entry["ccn_rank"].as_str().unwrap_or("");
        println!(
            "  {}  (callers={callers}, ccn={ccn} {rank})",
            entry["path"].as_str().unwrap_or("")
        );
        if let Some(shoulder) = entry["shoulder"].as_str() {
            if !shoulder.is_empty() {
                println!("    {shoulder}");
            }
        }
    }
    Ok(value)
}
